ts/ui/lazy/LazyHandler.ts

```diff
@@ -0,0 +1,483 @@
+      if (this.lazyTypeset) {
+        if (this.state() < STATE.TYPESET) {
+          const adaptor = document.adaptor;
+          if (!this.lazyMarker) {
+            const id = document.lazyList.add(this);
+            this.lazyMarker = adaptor.node('mjx-lazy', {[LAZYID]: id});
+            this.typesetRoot = adaptor.node('mjx-container', {}, [this.lazyMarker]);
+          }
+          this.state(STATE.TYPESET);
+        }
+      } else {
+        super.typeset(document);
+      }
+    }
+
+    /**
+     * When the MathItem is added to the page, set the observer to watch
+     *   for it coming into view so that it can be typeset.
+     *
+     * @override
+     */
+    public updateDocument(document: LazyMathDocument<N, T, D>) {
+      super.updateDocument(document);
+      if (this.lazyTypeset) {
+        document.lazyObserver.observe(this.lazyMarker as any as Element);
+      }
+    }
+
+    /**
+     * @override
+     */
+    public state(state: number = undefined, restore: boolean = false) {
+      //
+      // don't set the state if we are lazy processing
+      //
+      return (restore === null ? this._state : super.state(state, restore));
+    }
+
+  };
+
+}
+
+/*==========================================================================*/
+
+/**
+ * The properties added to MathDocument for lazy typesetting
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ */
+export interface LazyMathDocument<N, T, D> extends HTMLDocument<N, T, D> {
+
+  /**
+   * The Intersection Observer used to track the appearance of the expression markers
+   */
+  lazyObserver: IntersectionObserver;
+
+  /**
+   * The mapping of markers to MathItems
+   */
+  lazyList: LazyList<N, T, D>;
+
+}
+
+/**
+ * The mixin for adding lazy typesetting to MathDocuments
+ *
+ * @param {B} BaseDocument        The MathDocument class to be extended
+ * @return {LazyMathDocument}     The Lazy MathDocument class
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ * @template B  The MathDocument class to extend
+ */
+export function LazyMathDocumentMixin<N, T, D,
+B extends MathDocumentConstructor<HTMLDocument<N, T, D>>>(
+  BaseDocument: B
+): MathDocumentConstructor<HTMLDocument<N, T, D>> & B {
+
+  return class BaseClass extends BaseDocument {
+
+    /**
+     * The Intersection Observer used to track the appearance of the expression markers
+     */
+    public lazyObserver: IntersectionObserver;
+
+    /**
+     * The mapping of markers to MathItems
+     */
+    public lazyList: LazyList<N, T, D>;
+
+    /**
+     * A promise to make sure our compiling/typesetting is sequential
+     */
+    protected lazyPromise: Promise<void> = Promise.resolve();
+
+    /**
+     * The function used to typeset a set of lazy MathItems.
+     *   (uses requestIdleCallback if available, or setTimeout otherwise)
```

Comment:
Cool, `requestIdleCallback` is new to me!

Reply:
It was new to me when I was looking up how to detect elements that come into view.